playground/graph/graph.ml

```diff
@@ -0,0 +1,118 @@
+module Node : sig
+  type elt = int
+  type t
+  val create : elt -> t
+  val value : t -> elt
+  val compare : t -> t -> int
+  val id : t -> int
+  val to_string : t -> string
+end = struct
+  type elt = int
+  type t = { id : int; value : elt }
+  
+  let counter = ref 0
+
+  let create v = 
+    let id = !counter in
+    incr counter;
+    { id = id; value = v }
+
+  let value node = node.value
+
+  let id node = node.id
+  
+  let compare node1 node2 = Stdlib.compare node1.id node2.id
+
+  let to_string node = Printf.sprintf "Node(id: %d, value: %d)" node.id node.value
+end
+
+
+module NodeSet = Set.Make(Node)
+module NodeMap = Map.Make(Node)
+
+
+
+module Graph : sig
+  type elt = int  (* node values *)
+  type t    (* graphs *)
+  val empty : bool -> t
+  val add_node : Node.t -> t -> t
+  val remove_node : Node.t -> t -> t
+  val add_edge : Node.t -> Node.t -> t  -> t
+  val remove_edge : Node.t -> Node.t -> t -> t
+  val nodes : t -> Node.t List.t
+  val edges : t -> (Node.t * Node.t) list
+  val to_string : t -> string
+end = struct
+  type elt = int  (*Each node's values will be integers*)
+  type t = {
+    edges : NodeSet.t NodeMap.t; (*edges are represented with a Map whoose keys are of type Node.t and values are NodeSet.t*)
+    directed : bool;
+  }
+
+  let add_node graph node =
+    let edges = NodeMap.add node NodeSet.empty graph.edges in
+    { graph with edges }
+  
+  let remove_node node graph =
+    let edges =
+      graph.edges
+      |> NodeMap.remove node
+      |> NodeMap.map (NodeSet.remove node)
+    in
+    { graph with edges }
+
+  let add_edge node1 node2 graph =
+    let edges = graph.edges |> NodeMap.add node1 (NodeSet.add node2 (NodeMap.find node1 graph.edges)) in
```

Comment:
Ta del bi lahko dali v pomožno (in v signaturi skrito) funkcijo `add_undirected_edge`, ki doda samo enega, potem se pa v `add_edge` pokliče enkrat ali dvakrat. Podobno za `remove_edge`.

Reply:
7f3b11c4a2c4ec60ab3fa70b291a502b23cfc8a2